Extended Parameter DataSetMeasurements concept

RalphSteinhagen opened the pull request:
now includes also user-configurable
* DataSetMath measurements
* UTC/local-time/history trending of simple parameter values
* detaching of generated DataSets into new externally managed Stages
* added unit-tests for the corresponding FX UI items
* plus several minor bug fixes ...

Summary of additional changes
* HiddenSidesPane: respect requested StackPane Pos alignment requirements
* related: clean-up/simplification of Chart and ToolBarFlowPane
* LimitedIndexedTreeDataSet.java: added [set,get]SubtractOffset control
* ErrorDataSetRenderer: new parameter to control if all or only local DataSets shall be drawn
* fixed random JavaDoc errors detected with new IDE version

wirew0rm commented on chartfx-chart/src/main/java/de/gsi/chart/plugins/measurements/utils/DataSetSelector.java:
```diff
@@ -29,17 +31,32 @@ public DataSetSelector(final ParameterMeasurements plugin) {
-        if (!allDataSets.isEmpty()) {
-            selModel.select(0);
+        if (requiredNumberOfDataSets == 1) {
+            selModel.setSelectionMode(SelectionMode.SINGLE);
```
This doesn't seem to work as intended, I can e.g. select multiple datasets in the dialog for `DataSet1 - value`.

RalphSteinhagen replied:
fixed: superfluous line in DataSetMeasurements#initialize()...
removed: "SetListView().getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);"

RalphSteinhagen commented on chartfx-chart/src/main/java/de/gsi/chart/ui/ToolBarFlowPane.java:
```diff
@@ -68,13 +71,14 @@ private void adjustToolBarWidth() {
         double length = 0.0;
         for (Node node : this.getChildren()) {
-            length += node.prefWidth(-2);
+            length += node.prefWidth(28);
```
replaced by 'DEFAULT_TOOLBAR_HEIGHT' constant.

RalphSteinhagen commented on chartfx-chart/src/main/java/de/gsi/chart/viewer/DataViewWindow.java:
```diff
@@ -86,15 +88,17 @@
     private final ObjectProperty<WindowState> windowState = new SimpleObjectProperty<>(this, "windowState", WindowState.WINDOW_RESTORED) {
         @Override
         public void set(final WindowState state) {
+            // already armed, ignore redundant state changes
```
done

RalphSteinhagen commented on chartfx-samples/src/main/java/de/gsi/chart/samples/ErrorDataSetRendererSample.java:
```diff
@@ -136,8 +136,8 @@ public void start(final Stage primaryStage) {
         long startTime = ProcessingProfiler.getTimeStamp();
-        chart.getDatasets().add(dataSet);
-        chart.getDatasets().add(dataSetNoError);
+        chart.getRenderers().get(0).getDatasets().add(dataSet);
```
Yes, this is still needed if the computed DataSet measurement does not have the same x-Axis as the source DataSet (e.g. performing and plotting an FFT in the same chart). Issue lies in how the autorange axes are computed/updated in `Chart`. For most of the other measurement (especially DataSet1+-* [Value, DataSet2]) this isn't an issue. The problem would IMHO need to be addressed in another feature branch.

RalphSteinhagen commented on chartfx-chart/src/main/java/de/gsi/chart/viewer/DataViewWindow.java:
```diff
@@ -413,6 +431,18 @@ public BooleanProperty detachedProperty() {
     }
 
+    @Override
+    public boolean equals(Object obj) {
```
This is essentially needed that the same window/Pane/item is not added twice when switching from the detached to visibleChildren list. Added this, because sometimes the original DataViewWindow frame is being lost and regenerated when added back...

RalphSteinhagen commented on chartfx-chart/src/main/java/de/gsi/chart/plugins/measurements/DataSetMeasurements.java:
```diff
@@ -0,0 +1,871 @@
+                break;
+            case DIVIDE_VALUE:
+                FXUtils.runFX(() -> yAxis.set("(" + name1 + " / " + functionValue + ")", yAxisUnit));
+                outputDataSet.set(DataSetMath.mathFunction(firstDataSet, functionValue, MathOp.MULTIPLY));
```
fixed.

RalphSteinhagen commented on chartfx-chart/src/main/java/de/gsi/chart/plugins/measurements/DataSetMeasurements.java:
```diff
@@ -0,0 +1,871 @@
+                break;
+            case CONVERT_TO_DB:
+                FXUtils.runFX(() -> yAxis.set(MAG + name1 + ")", "dB"
+                                                                         + "(" + yAxisUnit + ")"));
```
opening bracket is coded in 'MAG?. avoided the +"(" + for brevity/speed.

wirew0rm replied:
I was referring to the code formatting, not the output formating, probably should have expressed that more clearly. I don't see why this is split on two lines.

RalphSteinhagen commented on chartfx-chart/src/main/java/de/gsi/chart/plugins/measurements/DataSetMeasurements.java:
```diff
@@ -0,0 +1,871 @@
+    public enum MeasurementType {
+        // basic math
+        ADD_FUNCTIONS(true, MATH, "DataSet1+DataSet2", 0, 2),
+        ADD_VALUE(true, MATH, "DataSet1 + value", 0, 1),
```
fixed

RalphSteinhagen commented:
> Looks very good, found only small bugs/oversights.
> 
> From trying out the samples:
> 
> * Chart anatomy sample: toolbar got huge
>   ![image](https://user-images.githubusercontent.com/1202371/78565991-e97f2f80-781e-11ea-8a17-2ddd5cd9c8c1.png)
> * Trending:
>   
>   * can select to trend itself, not really serious problem due to ratelimiter, but not really useful so it might be nicer to filter it out.
>     ![image](https://user-images.githubusercontent.com/1202371/78566804-0d8f4080-7820-11ea-8a49-6274192dd2bf.png)
>   * should probably show the title of the measurement instead of the dataset name
> * Reattaching measurements either by closing the detatched measurement chart or by `right-click -> dialog -> ok`  leads to missing axes for the measurement.
> * Performance:
>   
>   * ErrorDataSetRenderer 6% idle -> 40 - 140% with timer, skipping range updates -> 14% idle
> 
> Things probably unrelated to this PR:
> 
> * invert y axis on off toogle -> labels vanished (ErrorDataSetRendererStylingSample)
> * Batik cleaner thread? uses lots of cpu and does not terminate on its own.

* Chart anatomy sample: toolbar got huge
 ToolBarFlowPane is not properly managed by Parent (StackPane) and does not respect setWrapeLength(...) parameter... no idea (yet) of an easy workaround w/o rewriting more of  JavaFX 

* Trending:
Filtering implemented

* Reattaching measurements either [..]
occurs primarily when re-attaching/attaching -- now fixed

* Performance:
dropped axes update events are intentional and due to the very large number of data points (total: ~4e6 points)
